### ISSUE TITLE
MEMORY ISSUES in self-supervised.py

### ISSUE THREAD

**Issue by alexvaca0** [lucidrains/reformer-pytorch#42]
Hi there, I'm trying to pretrain a ReformerLM for spanish on a single Nvidia p-100 16GB GPU, and even when restricting the embedding dimension, the number of heads etc. I still get a Memory Error. I'm using the script in https://github.com/lucidrains/reformer-pytorch/blob/master/pretraining/self-supervised.py for that, and my configuration is the following: 

tokenizer.max_len=128
model = ReformerLM(
    num_tokens=tokenizer.vocab_size,
    dim=128,
    depth=1,
    heads=1,
    max_seq_len=tokenizer.max_len,
    causal=True
    n_hashes=2,
    ff_chunks=10000
)

trainer = ReformerTrainer(dataset, model, tokenizer, train_batch_size=1, eval_batch_size=1)

I've reduced the number of hashes, the max len, I've increased the ff_chunks... I've tried everything that's supposed to reduce the memory usage, but it's still not working. Have you been able to make the code in the link above work? @lucidrains If so, please tell me how... Just in case, my GPU is completely free before I start training, and the trainer tries to use about 19GB of memory...

**Comment by alexvaca0** [comment]
I think I already found the problem!! It was that   ReformerTrainer train() method, not only iterates through batches, but also for each data in that batch, and this causes the overflow. When eliminating that loop, the system works.

**Comment by lucidrains** [comment]
@alexvaca0 great that you found the problem! @zbloss  actually donated that script, so I'll circle him in so he knows!

**Comment by lucidrains** [comment]
@alexvaca0 could you send a pull request if you have the time?

**Comment by lucidrains** [comment]
@alexvaca0 also, im jealous of your GPU

**Comment by alexvaca0** [comment]
Perfect! I'll do it next Thursday for sure with the code fixed (that day is my I+Day, the day they let us work in innovative things at my company). If I find the time before that I'll do it sooner, I'm happy to contribute to your amazing work!! :) 
hahahah yeah it's a nice GPU actually. I wish I had one like that for my personal use, but I'm lucky for having access to it at my job.

**Comment by lucidrains** [comment]
Sounds good! I'll close this issue for now!
